libs/wire-subsystems/src/Wire/EnterpriseLoginSubsystem/Interpreter.hs

```diff
@@ -0,0 +1,321 @@
+    Member (Input (Maybe EnterpriseLoginSubsystemConfig)) r,
+    Member EmailSending r
+  ) =>
+  Domain ->
+  DomainRegistrationUpdate ->
+  Sem r ()
+updateDomainRegistrationImpl domain update = do
+  validate update
+  mOld <- tryGetDomainRegistrationImpl domain
+  let url =
+        "PUT /i/domain-registration/"
+          <> fromText (domainText domain)
+  new <- case mOld of
+    Just dr -> do
+      let new = dr {teamInvite = update.teamInvite, domainRedirect = update.domainRedirect} :: DomainRegistration
+      upsert $ toStored new
+      pure new
+    Nothing -> do
+      let dr = DomainRegistration domain update.domainRedirect update.teamInvite Nothing
+      upsert $ toStored dr
+      pure dr
+  sendAuditMail url "Domain registration updated" mOld (Just new)
+
+lockDomainImpl ::
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r,
+    Member (Input (Maybe EnterpriseLoginSubsystemConfig)) r,
+    Member EmailSending r
+  ) =>
+  Domain ->
+  Sem r ()
+lockDomainImpl domain = do
+  mOld <- tryGetDomainRegistrationImpl domain
+  let new = DomainRegistration domain Locked Allowed Nothing
+  upsert $ toStored new
+  let url =
+        "POST /i/domain-registration/"
+          <> fromText (domainText domain)
+          <> "/lock"
+  sendAuditMail url "Domain locked" mOld (Just new)
+
+unlockDomainImpl ::
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r,
+    Member (Input (Maybe EnterpriseLoginSubsystemConfig)) r,
+    Member EmailSending r
+  ) =>
+  Domain ->
+  Sem r ()
+unlockDomainImpl domain = do
+  old <- getDomainRegistrationImpl domain
+  let new = old {domainRedirect = None} :: DomainRegistration
+  case old.domainRedirect of
+    Locked -> upsert $ toStored new
+    _ -> throw EnterpriseLoginSubsystemUnlockError
+  let url =
+        "POST /i/domain-registration/"
+          <> fromText (domainText domain)
+          <> "/unlock"
+  sendAuditMail url "Domain locked" (Just old) (Just new)
+
+preAuthorizeImpl ::
+  ( Member DomainRegistrationStore r,
+    Member (Error EnterpriseLoginSubsystemError) r,
+    Member TinyLog r,
+    Member (Input (Maybe EnterpriseLoginSubsystemConfig)) r,
+    Member EmailSending r
+  ) =>
+  Domain ->
+  Sem r ()
+preAuthorizeImpl domain = do
+  mOld <- tryGetDomainRegistrationImpl domain
+  new <- case mOld of
+    Nothing -> do
+      let new = DomainRegistration domain PreAuthorized Allowed Nothing
+      upsert $ toStored new
+      pure new
+    Just sdr | sdr.domainRedirect == None -> do
+      let new = sdr {domainRedirect = PreAuthorized} :: DomainRegistration
+      upsert $ toStored new
+      pure new
+    Just sdr | sdr.domainRedirect == PreAuthorized -> pure sdr
+    _ -> throw EnterpriseLoginSubsystemErrorInvalidDomainRedirect
+  let url =
+        "POST /i/domain-registration/"
+          <> fromText (domainText domain)
+          <> "/preauthorize"
```

Comment:
it's not so nice that this url only accidentally coincides with the actual path of the handler that is eventually calling this action, but i don't have a better way.  :shrug:

Reply:
but in general: audit mails should be sent *before* the update, so we get false positives (bogus log entries), but not false negatives (missing log entries).

Reply:
> it's not so nice that this url only accidentally coincides with the actual path of the handler that is eventually calling this action, but i don't have a better way. 🤷

There is a better way. We can get it from the servant route type in the handler somehow (IIRC) and pass it all the way through. But I think it's not justified.